crates/dbs-miniball/README.md

```diff
@@ -1,21 +1,528 @@
+        and are required even with a single vCPU.
+
+        ```rust
+        // src/vm-vcpu/src/vm.rs
+        
+        #[cfg(target_arch = "x86_64")]
+        MpTable::new(vm.config.num_vcpus)?.write(guest_memory)?;
+        ```
+
+       2. Create KVM `irqchip`. This creates the virtual IOAPIC and virtual
+        PIC and sets up future vCPUs for local APIC.
+
+        ```rust
+        // src/vm-vcpu/src/vm.rs
+      
+        #[cfg(target_arch = "x86_64")]
+        vm.setup_irq_controller()?;
+        ```
+
+       3. Create vCPUs. An `fd` is registered with KVM for each vCPU.
+      
+        ```rust
+        // src/vm-vcpu/src/vm.rs
+        
+        vm.create_vcpus(bus, vcpus_config, guest_memory)?;
+        ```
+
+       4. Configure CPUID. Required (at least) because it’s the means by which
+        the guest finds out it’s virtualized.
+      
+        ```rust
+        // src/vm-vcpu/src/vcpu/mod.rs
+        
+        #[cfg(target_arch = "x86_64")]
+        let base_cpuid = _kvm
+        .get_supported_cpuid(kvm_bindings::KVM_MAX_CPUID_ENTRIES)
+        .map_err(Error::KvmIoctl)?;
+        ```
+
+       5. Configure MSRs (model specific registers). These registers control
+        (among others) the processor features. See the
+        [reference](https://www.intel.co.uk/content/dam/www/public/us/en/documents/manuals/64-ia-32-architectures-software-developer-system-programming-manual-325384.pdf#G14.8720).
+      
+        ```rust
+        // src/vm-vcpu/src/vcpu/mod.rs
+        #[cfg(target_arch = "x86_64")]
+        let supported_msrs = msrs::supported_guest_msrs(_kvm).map_err(Error::GetSupportedMsrs)?;
+        ```
+
+       6. Configure other registers (`kvm_regs`, `kvm_sregs`, `fpu`) and the
+        LAPICs.
+
+        ```rust
+        // src/vm-vcpu/src/vcpu/mod.rs
+      
+        #[cfg(target_arch = "x86_64")]
+        {
+            vcpu.configure_cpuid(&vcpu.config.cpuid)?;
+            vcpu.configure_msrs()?;
+            vcpu.configure_sregs(memory)?;
+            vcpu.configure_lapic()?;
+            vcpu.configure_fpu()?;
+        }
+        ```
+
+
+
+5. Configure legacy devices. This is done partially through `kvm-ioctls`,
+   partially (serial console emulation) through `vm-superio`. Device event
+   handling is mediated with `event-manager`.
+   1. Requirements: KVM is configured, guest memory is configured, `irqchip`
+      is configured (`x86_64`), event manager is configured
+   2. Inputs: N/A
+   3. Breakdown:
+      1. Create dummy speaker. This needs to be emulated because some kernels
+        access the speaker’s port, leading to continuous KVM exits for the
+        otherwise unhandled device IO.
+      2. Create serial console.
+
+```rust
+// src/vmm/src/lib.rs
+        
+vmm.add_serial_console()?;
+#[cfg(target_arch = "x86_64")]
+vmm.add_i8042_device()?;
+```
+
+6. Configure root block device. This is done through `vm-virtio`. Device event
```

Comment:
[dbs-virtio-devices](https://github.com/openanolis/dragonball-sandbox/tree/main/crates/dbs-virtio-devices) could be useful .